flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java

```diff
@@ -0,0 +1,179 @@
+        }
+
+        return new DefaultLeaderElectionService(
+                multipleComponentLeaderElectionService.createDriverFactory(leaderName));
+    }
+
+    @GuardedBy("lock")
+    private MultipleComponentLeaderElectionService getOrInitializeSingleLeaderElectionService() {
+        if (multipleComponentLeaderElectionService == null) {
+            try {
+                multipleComponentLeaderElectionService =
+                        new DefaultMultipleComponentLeaderElectionService(
+                                fatalErrorHandler,
+                                "Single leader election service.",
+                                new ZooKeeperMultipleComponentLeaderElectionDriverFactory(
+                                        leaderNamespacedCuratorFramework));
+            } catch (Exception e) {
+                throw new FlinkRuntimeException(
+                        String.format(
+                                "Could not initialize the %s",
+                                DefaultMultipleComponentLeaderElectionService.class
+                                        .getSimpleName()),
+                        e);
+            }
+        }
+
+        return multipleComponentLeaderElectionService;
+    }
+
+    @Override
+    protected LeaderRetrievalService createLeaderRetrievalService(String leaderPath) {
+        // Maybe use a single service for leader retrieval
+        return ZooKeeperUtils.createLeaderRetrievalService(
+                leaderNamespacedCuratorFramework, leaderPath, configuration);
+    }
+
+    @Override
+    protected void internalClose() throws Exception {
+        Exception exception = null;
+        synchronized (lock) {
+            if (multipleComponentLeaderElectionService != null) {
+                try {
+                    multipleComponentLeaderElectionService.close();
+                } catch (Exception e) {
+                    exception = e;
+                }
+                multipleComponentLeaderElectionService = null;
+            }
+        }
+
+        try {
+            super.internalClose();
+        } catch (Exception e) {
+            exception = ExceptionUtils.firstOrSuppressed(e, exception);
+        }
+
+        ExceptionUtils.tryRethrowException(exception);
+    }
+
+    @Override
+    protected void internalCleanupJobData(JobID jobID) throws Exception {
+        super.internalCleanupJobData(jobID);
+    }
+
+    @Override
+    protected String getLeaderPathForResourceManager() {
+        return ZooKeeperUtils.getResourceManagerNode();
+    }
+
+    @Override
+    protected String getLeaderPathForDispatcher() {
+        return ZooKeeperUtils.getDispatcherNode();
+    }
+
+    @Override
+    protected String getLeaderPathForJobManager(JobID jobID) {
+        return jobID.toString();
+    }
+
+    @Override
+    protected String getLeaderPathForRestServer() {
+        return ZooKeeperUtils.getRestServerNode();
+    }
```

Comment:
I think you are right that there is a difference between the method name and what is returned. In this case I would suggest to change the naming of the methods to `getLeaderNameXYZ` instead of `getLeaderPathXZY`.

Reply:
Arg, the problem is that the different `AbstractHaServices` implementations treat this method differently. Since I don't want to touch the `ZooKeeperHaServices`, I will keep it as is. Once `ZooKeeperHaServices` is removed, we can clean it up.